src/BoltOn.Data.CosmosDb/BaseCosmosDbRepository.cs

```diff
@@ -0,0 +1,125 @@
+        {
+            var query = _client.CreateDocumentQuery<TEntity>(GetDocumentCollectionUri(), FeedOptions)
+                .Where(predicate)
+                .AsDocumentQuery();
+
+            return GetResultsFromDocumentQuery(query).Result;
+        }
+
+        public virtual async Task<IEnumerable<TEntity>> FindByAsync(Expression<Func<TEntity, bool>> predicate,
+            CancellationToken cancellationToken = default(CancellationToken), params Expression<Func<TEntity, object>>[] includes)
+        {
+            var query = _client.CreateDocumentQuery<TEntity>(GetDocumentCollectionUri(), FeedOptions)
+               .Where(predicate)
+               .AsDocumentQuery();
+
+            return await GetResultsFromDocumentQuery(query);
+        }
+
+        public virtual IEnumerable<TEntity> GetAll()
+        {
+            return _client.CreateDocumentQuery<TEntity>
+                (GetDocumentCollectionUri(), FeedOptions)
+                .AsEnumerable();
+        }
+
+        public virtual Task<IEnumerable<TEntity>> GetAllAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return Task.FromResult(_client.CreateDocumentQuery<TEntity>
+                (GetDocumentCollectionUri(), FeedOptions)
+                .AsEnumerable());
+        }
+
+        public virtual TEntity GetById<TId>(TId id)
+        {
+            var document = _client.ReadDocumentAsync<TEntity>(GetDocumentUri(id.ToString()), RequestOptions).Result;
+            return document.Document;
+        }
+
+        public virtual async Task<TEntity> GetByIdAsync<TId>(TId id)
+        {
+            var document = await _client.ReadDocumentAsync<TEntity>(GetDocumentUri(id.ToString()), RequestOptions);
+            return document.Document;
+        }
+
+        public virtual void Update(TEntity entity)
+        {
+            _client.UpsertDocumentAsync(GetDocumentCollectionUri(), entity, RequestOptions).Wait();
+        }
+
+        public virtual async Task UpdateAsync(TEntity entity, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await _client.UpsertDocumentAsync(GetDocumentCollectionUri(), entity, RequestOptions);
+        }
+
+        protected Uri GetDocumentCollectionUri()
+        {
+            return UriFactory.CreateDocumentCollectionUri(_databaseName, _collectionName);
+        }
+
+        protected Uri GetDocumentUri(string id)
+        {
+            return UriFactory.CreateDocumentUri(_databaseName, _collectionName, id);
+        }
+
+        protected async Task<IEnumerable<TEntity>> GetResultsFromDocumentQuery(IDocumentQuery<TEntity> query)
+        {
+            List<TEntity> results = new List<TEntity>();
+            while (query.HasMoreResults)
+            {
+                results.AddRange(await query.ExecuteNextAsync<TEntity>());
+            }
+            return results;
+        }
+
+        public void Init<TOptions>(TOptions options)
```

Comment:
what if make the consumers to override the repository methods in case if they have to set the RequestOptions and FeedOptions i.e., not allow them to set these values outside the repos